File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <unordered_map>

using namespace std;

int visited[11];
int check[11];
unordered_map<string,int>mp;

void bt(int d,string s){
    for(int i=d;i<s.size();i++){
        if(visited[i]==0){
            visited[i]=1;
            bt(i+1,s);
            visited[i]=0;
        }
    }
    string st = "";
    for(int i=0;i<s.size();i++){
        if(visited[i]==1){
            st+=s[i];
        }
    }
    if(check[st.size()]==1){
        mp[st]++;
    }
    
}

vector<string> solution(vector<string> orders, vector<int> course) {
    vector<string> answer;
    
    for(int c : course){
        check[c]=1;
    }
    
    for(string &s : orders){
        sort(s.begin(),s.end());
        bt(0,s);
    }
    
    for(int c : course){
        int maxOrder = 0;
        for(pair<string,int> e : mp){
            if(e.first.size()==c)
                maxOrder = max(maxOrder,e.second);
        }
        if(maxOrder <= 1) continue;
        for(pair<string,int> e : mp){
            if(maxOrder == e.second && e.first.size()==c){
                answer.push_back(e.first);
            }
        }
    }
    sort(answer.begin(),answer.end());// 사전순으로 정렬
    
    
    return answer;
}

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.*;

class Solution {
    private static HashMap<String, Integer> hm = new HashMap<>();
    private static HashSet<Integer> hs = new HashSet<>();
    
    public static void combinations(int idx, char[] order, String result){
        if(hs.contains(result.length())){
            hm.put(result, hm.getOrDefault(result,0)+1);
        }
        
        for(int i = idx ; i<order.length;i++){
            combinations(i+1, order, result + order[i]);
        }
        
    }
    
    public String[] solution(String[] orders, int[] course) {
        
        for(int c : course){
            hs.add(c);
        }
        
        for(String o : orders){
            char[] orderArray = o.toCharArray();
            Arrays.sort(orderArray);
            combinations(0, orderArray, "");
        }
        
        // 코스 길이별 최대 빈도 저장
        HashMap<Integer, Integer> maxCount = new HashMap<>();
        
        for(String key : hm.keySet()){
            int cnt = hm.get(key);
            int len = key.length();
            
            if(cnt < 2) continue;
            
            maxCount.put(len, Math.max(maxCount.getOrDefault(len,0),cnt));
        }
        
        List<String> answer = new ArrayList<>();
        
        for(String key : hm.keySet()){
            int len = key.length();
            int cnt = hm.get(key);
            
            if(cnt>=2 && maxCount.getOrDefault(len,0) ==cnt){
                answer.add(key);
            }
        }
        
        Collections.sort(answer);
        
        
        return answer.toArray(new String[0]);
    }
}